kernel/cpu_x86_64.c

```diff
@@ -0,0 +1,196 @@
+    FILL_TRAP_GATE(3, 2);
+    FILL_TRAP_GATE(4, 2);
+    FILL_TRAP_GATE(5, 2);
+    FILL_TRAP_GATE(6, 2);
+    FILL_TRAP_GATE(7, 2);
+    FILL_TRAP_GATE(8, 3); /* #DF runs on IST3 (cpu_nmi_stack) */
+    FILL_TRAP_GATE(10, 2);
+    FILL_TRAP_GATE(11, 2);
+    FILL_TRAP_GATE(12, 2);
+    FILL_TRAP_GATE(13, 2);
+    FILL_TRAP_GATE(14, 2);
+    FILL_TRAP_GATE(16, 2);
+    FILL_TRAP_GATE(17, 2);
+    FILL_TRAP_GATE(18, 2);
+    FILL_TRAP_GATE(19, 2);
+    FILL_TRAP_GATE(20, 2);
+
+    /*
+     * Load irq vectors. All irqs run on IST1 (cpu_intr_stack).
+     */
+#define FILL_IRQ_GATE(num, ist) extern void cpu_irq_##num(void); \
+    idt_fillgate(32 + num, cpu_irq_##num, ist)
+    FILL_IRQ_GATE(0, 1);
+    FILL_IRQ_GATE(1, 1);
+    FILL_IRQ_GATE(2, 1);
+    FILL_IRQ_GATE(3, 1);
+    FILL_IRQ_GATE(4, 1);
+    FILL_IRQ_GATE(5, 1);
+    FILL_IRQ_GATE(6, 1);
+    FILL_IRQ_GATE(7, 1);
+    FILL_IRQ_GATE(8, 1);
+    FILL_IRQ_GATE(9, 1);
+    FILL_IRQ_GATE(10, 1);
+    FILL_IRQ_GATE(11, 1);
+    FILL_IRQ_GATE(12, 1);
+    FILL_IRQ_GATE(13, 1);
+    FILL_IRQ_GATE(14, 1);
+    FILL_IRQ_GATE(15, 1);
+
+    volatile struct idtptr idtptr;
+    idtptr.limit = sizeof(cpu_idt) - 1;
+    idtptr.base = (uint64_t) &cpu_idt;
+    __asm__ __volatile__("lidt (%0)" :: "r" (&idtptr));
+}
+
+static struct tss cpu_tss;
+
+static char cpu_intr_stack[4096]; /* IST1 */
+static char cpu_trap_stack[4096]; /* IST2 */
+static char cpu_nmi_stack[4096];  /* IST3 */
+
+static void tss_init(void)
+{
+    extern uint64_t cpu_gdt64[];
+    struct tss_desc *td = (void *)&cpu_gdt64[GDT_DESC_TSS_LO];
+
+    cpu_tss.ist[0] = (uint64_t)&cpu_intr_stack[sizeof cpu_intr_stack];
+    cpu_tss.ist[1] = (uint64_t)&cpu_trap_stack[sizeof cpu_trap_stack];
+    cpu_tss.ist[2] = (uint64_t)&cpu_nmi_stack[sizeof cpu_nmi_stack];
+    td->limit_lo = sizeof(cpu_tss);
+    td->base_lo = (uint64_t)&cpu_tss;
+    td->type = 0x9;
+    td->dpl = 0;
+    td->p = 1;
+    td->limit_hi = 0;
+    td->gran = 0;
+    td->base_hi = (uint64_t)&cpu_tss >> 24;
+    td->zero = 0;
+
+    cc_barrier();
+    __asm__ __volatile__("ltr %0" :: "r" ((unsigned short)(GDT_DESC_TSS_LO * 8)));
+}
+
+void cpu_init(void)
+{
+    gdt_init();
+    tss_init();
+    idt_init();
+    platform_intr_init();
+}
+
+static char *traps[32] = {
+    "#DE", "#DB", "#NMI", "#BP", "#OF", "#BR", "#UD", "#NM", "#DF", "#9", "#TS",
+    "#NP", "#SS", "#GP", "#PF", "#15", "#MF", "#AC", "#MC", "#XM", "#VE", "#21",
+    "#22", "#23", "#24", "#25", "#26", "#27", "#28", "#29", "#30", "#31"
+};
+
+void trap_handler(uint64_t num, struct trap_regs *regs)
+{
+    log(INFO, "Solo5: trap: type=%s ec=0x%lx rip=0x%lx rsp=0x%lx rflags=0x%lx\n",
+        traps[num], regs->ec, regs->rip, regs->rsp, regs->rflags);
+    if (num == 14)
+        log(INFO, "Solo5: trap: cr2=0x%lx\n", regs->cr2);
+    PANIC("Fatal trap");
+}
+
+/* keeps track of how many stacked "interrupts_disable"'s there are */
+int intr_depth = 1;
```

Comment:
Sorry I misread, I wrongly had the impression that the variable was also accessed when handling exceptions/traps.